autonav_ws/src/autonav_odometry/src/main.py

```diff
@@ -0,0 +1,113 @@
+        # init filters
+        # IF WE IMPLEMENT MORE FILTERS ADD THEM HERE (AND ANYWHERE THAT COMES UP WHEN YOU CTRL-F "NEW FILTERS")!
+        self.pf = ParticleFilter(self.latitude_length, self.longitude_length)
+        self.config = FiltersNodeConfig()
+        self.onReset()
+
+    def config_updated(self, jsonObject):
+        self.config = json.loads(self.jdump(jsonObject), object_hook=lambda d: SimpleNamespace(**d))
+        # reinit filters after getting new config
+        self.onReset()
+
+    def init(self):
+        # [melee announcer voice] ready...
+        self.create_subscription(GPSFeedback, "/autonav/gps", self.onGPSReceived, 20)
+        self.create_subscription(MotorFeedback, "/autonav/MotorFeedback", self.onMotorFeedbackReceived, 20)
+        self.position_publisher = self.create_publisher(Position, "/autonav/position", 20)
+        # go!
+        self.set_device_state(DeviceState.OPERATING)
+
+    def onReset(self):
+        self.first_gps = None
+        self.last_gps = None
+        self.pf.init_particles()
+        # INIT NEW FILTERS HERE
+
+    def system_state_transition(self, old: SystemState, updated: SystemState):
+        # reinit filters when robot changes modes
+        if old.state != SystemState.AUTONOMOUS and updated.state == SystemState.AUTONOMOUS:
+            self.onReset()
+
+        if old.state != SystemState.MANUAL and updated.state == SystemState.MANUAL:
+            self.onReset()
+        # or gets going
+        if old.mobility == False and updated.mobility == True:
+            self.onReset()
+
+    def onGPSReceived(self, msg: GPSFeedback):
+        if msg.gps_fix == 0 and msg.is_locked == False:
+            return
+
+        if self.first_gps is None:
+            self.first_gps = msg
+
+        self.last_gps = msg
+        # USE NEW FILTERS HERE
+        if self.config.filter_type == FilterType.PARTICLE_FILTER:
+            gps_x, gps_y = self.pf.gps(msg)
+            self.log(f"PF back from GPS ping w/ X={gps_x}, Y={gps_y}", LogLevel.DEBUG)
+        else:
+            self.log(f"{self.config.filter_type} isn't a valid FilterType! Did you implement a new filter and forget to use it?", LogLevel.FATAL)
+
+    def onMotorFeedbackReceived(self, msg: MotorFeedback):
+        averages = None
+        # USE NEW FILTERS HERE
+        if self.config.filter_type == FilterType.PARTICLE_FILTER:
+            avg_x, avg_y, avg_theta = self.pf.feedback(msg)
+            self.log(f"PF back from motor odom w/ X={avg_x}, Y={avg_y}, θ={avg_theta}", LogLevel.DEBUG)
+        else:
+            self.log(f"{self.config.filter_type} isn't a valid FilterType! Did you implement a new filter and forget to use it?", LogLevel.FATAL)
+
+        if averages is None:
+            return
+        # unpack what the filter just returned
+        position = Position()
+        position.x = averages[0]
+        position.y = averages[1]
+        position.theta = averages[2]
```

Comment:
if you wanted to you could do `position.x, position.y, position.theta = averages` but I don't really care either way, just plugging some Python tricks